pkg/tools/etcd_helper.go

```diff
@@ -99,6 +99,11 @@ func (h *EtcdHelper) listEtcdNode(key string) ([]*etcd.Node, uint64, error) {
 			return nodes, index, err
 		}
 	}
+	if result.Node.Value != "" {
```

Comment:
Now that I take a second look, I'm a little concerned by this approach of modifying the data etcd sends back. I would feel much more comfortable if instead you made ListPredicate do an ExtractObj + package into a list (or maybe better, write a helper function here that does that).

I guess I'm just nervous that genuine listing will break if a parent node somehow gets a value, plus vague unquantifiable worries that etcd will change something and this will break...

Does this make sense? Can you either follow this suggestion or add a comment before this if statement that explains why you're monkeying with etcd's returned data?